### ISSUE TITLE
Negative value specified for TestSessionTimeout - Failed to create a TestRunner, Stryker cannot recover from this

### ISSUE THREAD

**Issue by GrahamTheCoder** [stryker-mutator/stryker-net#2069]
**Describe the bug**

I got some warnings like this: `Runner 1: Some mutations were executed outside any test`
It then looks like it got a bit of the way through (2.47%) then blew up with:
```csharp
*** Failed to create a TestRunner, Stryker cannot recover from this!***


   --- End of inner exception stack trace ---
   at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection exceptions, CancellationToken cancelToken, Exception otherException)
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IList`1 list, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body)
   at Stryker.Core.MutationTest.MutationTestProcess.TestMutants(IEnumerable`1 mutantsToTest) in /_/src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs:line 147
   at Stryker.Core.MutationTest.MutationTestProcess.Test(IEnumerable`1 mutantsToTest) in /_/src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs:line 131
   at Stryker.Core.StrykerRunner.RunMutationTest(IStrykerInputs inputs, ILoggerFactory loggerFactory, IProjectOrchestrator projectOrchestrator) in /_/src/Stryker.Core/Stryker.Core/StrykerRunner.cs:line 107
   at Stryker.CLI.StrykerCli.RunStryker(IStrykerInputs inputs) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 93
   at Stryker.CLI.StrykerCli.<>c__DisplayClass10_0.<Run>b__0() in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 68
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.<>c__DisplayClass143_0.<OnExecute>b__0(CancellationToken _)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.ExecuteAsync(String[] args, CancellationToken cancellationToken)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at Stryker.CLI.StrykerCli.Run(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 74
   at Stryker.CLI.Program.Main(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/Program.cs:line 14
```

**Logs**
A huge amount of stuff is in those logs. I'd have to look through more carefully before just uploading them. The phrase "a TestRunner" doesn't appear in any of them, so I'm not sure how much they'll help.

**Expected behavior**
Keeps running to completion. Exits with output based on tests successfully completed.

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Type of project: Net 6
 - Stryker Version: 1.5.3
 - Runtime/sdk version: 6.0.300-preview.22154.4

**Additional context**

Thanks for all the work on this project, I've been trying it occasionally over the years and this is by far the easiest to setup and closest I've got it to working yet.
Sorry if there's a doc link for this sort of error, Google didn't find it 

##### Probably a separate issue

For my use case, even if there is some issue, ideally it'd keep going. Luckily I ran it restricted to changes since a recent commit, and checked first, but I was going to leave it running overnight since it predicted 5 hours for the full run, and it'd be a bit sad if it failed after 4 hours with absolutely no output. Is there some way to do incremental output, or automatically/manually "run again, but skip the thing that blew up last time"?

**Comment by rouke-broersma** [comment]
Just to be sure have you tried running stryker on a non-preview version of the dotnet sdk?

**Comment by GrahamTheCoder** [comment]
Thanks both. I've just realized my file search was filtered to the wrong filetype, and now I can see
```
2022-05-27T10:34:29.8946227+01:00  [VRB] Runner 5: [Error] Microsoft.VisualStudio.TestPlatform.ObjectModel.SettingsException: Invalid settings 'RunConfiguration'.  Invalid value '-34119' specified for 'TestSessionTimeout'.
	Line 68741: System.Exception: Invalid settings 'RunConfiguration'.  Invalid value '-34119' specified for 'TestSessionTimeout'.
```
I'll try a really big additional timeout to counterract that

**Comment by GrahamTheCoder** [comment]
In Azure DevOps I've also seen some very odd test durations coming out recently for some other tests. I wonder if there's an issue in some test runner/framework code.

In terms of unblocking the use of this tool, would you accept a PR that logs a warning if it's negative and sets the value calculated here to some default value (e.g. 0):
https://github.com/stryker-mutator/stryker-net/blob/c19bc08ca4533d3f2dd4f84514e92c9d093ce1ad/src/Stryker.Core/Stryker.Core/Initialisation/InitialTestProcess.cs#L60-L62

**Comment by GrahamTheCoder** [comment]
> I would love to see how you tried to fix this.

It was literally just taking away the `if xUnit` and using its implementation in both cases. i.e. `_initialResults.FirstOrDefault()?.Duration ?? TimeSpan.Zero`
I didn't have time to delve much into the structure of the object (i.e. whether it even had multiple results - I think you're saying it doesn't). I think it had a native object wrapped so was annoying to debug. I can probably try to serialize as JSON though and see if there's an obvious pattern

The test project in question uses 3.13.3 which looks like the latest. I'm not sure which other versions of things might make a different. e.g. Could either of these two be interfering?
```
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.7.1" />
    <PackageReference Include="NUnit3TestAdapter" Version="4.0.0" />
```

> more along the lines of adding a CLI options that would disregard errors and attempt to go through

Yeah, I understand that knowing about individual different issues is important and things that don't cause breakage tend to go unreported. I thought of the command line option initially, but it seemed like quite a bit of work to still get what's still not very good solution: e.g. User experiences potentially transient failure -> read the docs -> tendency to use the option everywhere from now on, still masking the issues.

If there are a bunch of other known issues that stop the system in its tracks, then it may make more sense to actually have it carry on by default but add a section of the output report dedicated to "These 20% of your tests could not be analysed" due to the following list of issues, please report them for more accurate results in future". That would work nicely with the current safe mode + warnings and dev-mode options I've seen for other issues such as https://github.com/stryker-mutator/stryker-net/issues/1923 which I see you're looking to improve here: https://github.com/stryker-mutator/stryker-net/issues/1953

So I haven't read much about it, but I think I'm proposing to allow safe mode to apply to a test (perhaps now it's just to source files?) and could open an issue for that?
Separately, it might be worth opening an issue to improve the display of issues and their impact to increase reporting

Tomorrow I'll try to debug the main issue here again if I get the chance

**Comment by dupdob** [comment]
>Yeah, I understand that knowing about individual different issues is important and things that don't cause breakage tend to go unreported. I thought of the command line option initially, but it seemed like quite a bit of work to still get what's still not very good solution: e.g. User experiences potentially transient failure -> read the docs -> tendency to use the option everywhere from now on, still masking the issues.

That may be a risk, but I fail to see how hiding it preemptively is better. And there are strategies to prevent this kind of overgeneralization. But I suggest it would better to discuss in a separate topic and stay on track here.

Any chance I can access this project somehow ? At this stage this is either a new issue in timeout calculation in Stryker (as when we discovered the hard way how xUnit was lying about it), or that NUnit duration reports are invalid. But is seems hard to diagnose without a reproduction scenario

**Comment by GrahamTheCoder** [comment]
I'm afraid I can't share the project source code, binaries or anything useful.

Not sure you can see clearly on the screenshot, but it is essentially that the NUnit duration reports are invalid...or at least unexpected.
To clarify: The stopwatch (and my own estimation) is 22 seconds. The sum of the NUnit test durations is 71 seconds.

The reason I suspect setup time being included in each one, is that I have a large number of *very* fast dynamically generated test cases. I'll try to dig into why that is tomorrow. Perhaps I'll be able to cutdown the repro until it's something shareable.

**Comment by GrahamTheCoder** [comment]
Looking at the testResults object from here:
https://github.com/stryker-mutator/stryker-net/blob/46985134192ade92d5e56dc2ce45681ff92e41c9/src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs#L200

Here's a cut down example of two (out of ~1000) subsequent test cases from the same NUnit TestCaseSource attribute. Each test case is listed separately but they all have quite similar durations and heavily overlap in start/end time:
```json
...
{
    "TestCase": {
        "Id": "eb223fbf-0031-6b4e-f051-c17dbafb4eba",
    },
    "Duration": "00:00:00.1245178",
    "StartTime": "2022-06-06T11:07:32.3340251+00:00",
    "EndTime": "2022-06-06T11:07:32.4585429+00:00"
},
{
    "TestCase": {
        "Id": "8d328e4a-395c-417f-8770-ee4b5f43fc28"
    },
    "Duration": "00:00:00.1242651",
    "StartTime": "2022-06-06T11:07:32.3351958+00:00",
    "EndTime": "2022-06-06T11:07:32.4594609+00:00"
}
...
```

**Comment by GrahamTheCoder** [comment]
I've fixed the issue in my project's tests leading to duplicate names so I'm now happily on my way to a full mutation report 🥳

> probably a mix of duration aggregation and start/end time computation

Yeah makes sense. If you aren't sure you can trust the Duration property, you could always aggregate the `EndTime-StartTime` *per test case* like this to fix the issue:
`acc.Duration += next.EndTime - next.StartTime;`

**Comment by dupdob** [comment]
While redesigning the code, I encountered several issues, which were hidden being the duration re computation logic.
Having added dedicated unit tests, I am confident we can reach a situation where we trustfully process the time reported by the runners.

**Comment by rouke-broersma** [comment]
@GrahamTheCoder @dupdob Is this still a problem we need to look into?

**Comment by GrahamTheCoder** [comment]
#2039 was intended to fix it. There's a repro above if someone wants to retest on a particular version
https://github.com/stryker-mutator/stryker-net/issues/2069#issuecomment-1206115990

**Comment by rouke-broersma** [comment]
I am not observing any failures or errors in the testrunners when running against your reproduction and I am getting timeouts on some static mutants (probably due to changed timeout average with these new randomly generated tests) so I am gonna go ahead and mark this fixed for now :)
